protocol/encoder/encoder.go

```diff
@@ -0,0 +1,82 @@
+package encoder
+
+import (
+	"encoding/binary"
+)
+
+type Encoder struct {
+	bytes  []byte
+	offset int
+}
+
+func NewEncoder(bytes []byte) *Encoder {
+	return &Encoder{
+		bytes:  bytes,
+		offset: 0,
+	}
+}
+
+func (re *Encoder) Offset() int {
+	return re.offset
+}
+
+func (re *Encoder) AllBytes() []byte {
```

Comment:
The interface for this could be far simpler. I wouldn't expose `Offset` and `AllBytes` - no caller is going to require these values, they'll only need the "written" bytes, and doing `len(bytes)` will give them the offset. 

Would remove these both and make `EncodedBytes` just `Bytes`.